### PR TITLE
Summary openai supabase

### PR DESCRIPTION
**Pull request by htahir1**
WIP 

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
Briefly describe the changes you have introduced.

### CODE REVIEW

File: _posts/2023-04-30-openai-supabase-summary.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+tags: tooling zenml evergreen llm foundationmodels
+publish_date: April 30, 2023
+date: 2023-04-30T00:02:00Z
+thumbnail: /assets/posts/gpt-4-supabase-summary/youtldr_summarizer_slack.png
+image:
+  path: /assets/posts/gpt-4-supabase-summary/youtldr_summarizer_slack.png
+---
+
+**Last updated:** April 30, 2023
+# Using ZenML with GPT-4 to Analyze and Version Data from Supabase: A Case Study with you-tldr.com
```
Comment (strickvl) [inline _posts/2023-04-30-openai-supabase-summary.md:16]:
```suggestion
```

Title is already populated above.

File: _posts/2023-04-30-openai-supabase-summary.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+In this blog post, we'll explore how ZenML can be used in conjunction with OpenAI's GPT-4 to analyze and version data from a Supabase database. We'll use the you-tldr.com website as an example, showcasing how the site populates an analytics table in Supabase and how ZenML asynchronously picks up the latest video data for analysis.
```
Comment (strickvl) [inline _posts/2023-04-30-openai-supabase-summary.md:18]:
```suggestion
In this blog post, we'll explore how ZenML can be used in conjunction with OpenAI's GPT-4 to analyze and version data from a Supabase database. We'll use the [you-tldr.com](https://you-tldr.com/) website as an example, showcasing how the site populates an analytics table in Supabase and how ZenML asynchronously picks up the latest video data for analysis.
```

File: _posts/2023-04-30-openai-supabase-summary.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+Large language models (LLMs) like GPT-4 have revolutionized natural language processing, offering unparalleled capabilities for knowledge generation and reasoning. However, incorporating custom, private data into these models remains a challenge. ZenML, an extensible, open-source MLOps framework, can help overcome this limitation by versioning data and allowing for comparisons between summaries rather than raw data.
+
+# 🧑‍🎓 The you-tldr.com Case Study
+
+[you-tldr.com](https://you-tldr.com) is a website that provides concise summaries of YouTube videos. The site populates an analytics table in Supabase with the latest video data from its users. We'll demonstrate how ZenML can be used to analyze this data using GPT-4 and version the summaries for comparison.
```
Comment (strickvl) [inline _posts/2023-04-30-openai-supabase-summary.md:24]:
```suggestion
[you-tldr.com](https://you-tldr.com) is a website that provides concise summaries of YouTube videos. The site populates an analytics table in Supabase with information about the kinds of videos that users are choosing to summarize. We'll demonstrate how ZenML can be used to analyze this data using GPT-4 and version the summaries for comparison.
```

File: _posts/2023-04-30-openai-supabase-summary.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+| 'A Bicycle of the Mind' - Steve Jobs on the Computer                                                 | 2023-04-11 13:47:25.8898+00   |
+| 'A Day in the Life' at My Off Grid Property! | Fish Pond, Forest Maintenance, Winching, Sawmill Shed | 2023-04-29 20:34:57.28206+00  |
+| 'A Tale of Two Beasts' read by Sarah Silverman                                                       | 2023-03-23 13:50:36.405557+00 |
+| 'Brave Irene' read by Al Gore                                                                        | 2023-03-29 02:40:08.482399+00 |
+| 'Brilliant' and 'disgusting': New Trump ad pokes fun at DeSantis' alleged eating habit               | 2023-04-18 01:45:52.157644+00 |
+| 'Censorship we can get behind': Elon Musk tears down tweet celebrating 'paedophile pride day'        | 2023-04-30 00:34:06.551438+00 |
+| 'Clark the Shark' read by Chris Pine                                                                 | 2023-03-27 02:49:25.704303+00 |
+| 'Emergency' Fed rate cut by June, only 6 U.S. banks will be left by 2025 paving way for CBDC - Dowd  | 2023-03-25 09:02:01.49922+00  |
+
```
Comment (strickvl) [inline _posts/2023-04-30-openai-supabase-summary.md:46]:
```suggestion

```
| video_title                                                                                          | created_at                    |
| ---------------------------------------------------------------------------------------------------- | ----------------------------- |
| $75,000 Grants Now Available 10,000 Grants - Startups - Existing - Self Employed FREE MONEY NOT LOAN | 2023-04-19 16:39:23.157915+00 |
| $90 3v3 Challenge: Domi, Size, Hummi vs Crazy, Ammo, Boycah | C&C Generals Zero Hour                 | 2023-04-30 00:29:03.918995+00 |
| $9000 Table Build                                                                                    |
| 'A Bad Case of Stripes' read by Sean Astin                                                           | 2023-03-29 08:43:27.115494+00 |
| 'A Bicycle of the Mind' - Steve Jobs on the Computer                                                 | 2023-04-11 13:47:25.8898+00   |
| 'A Day in the Life' at My Off Grid Property! | Fish Pond, Forest Maintenance, Winching, Sawmill Shed | 2023-04-29 20:34:57.28206+00  |
| 'A Tale of Two Beasts' read by Sarah Silverman                                                       | 2023-03-23 13:50:36.405557+00 |
| 'Brave Irene' read by Al Gore                                                                        | 2023-03-29 02:40:08.482399+00 |
| 'Brilliant' and 'disgusting': New Trump ad pokes fun at DeSantis' alleged eating habit               | 2023-04-18 01:45:52.157644+00 |
| 'Censorship we can get behind': Elon Musk tears down tweet celebrating 'paedophile pride day'        | 2023-04-30 00:34:06.551438+00 |
| 'Clark the Shark' read by Chris Pine                                                                 | 2023-03-27 02:49:25.704303+00 |
| 'Emergency' Fed rate cut by June, only 6 U.S. banks will be left by 2025 paving way for CBDC - Dowd  | 2023-03-25 09:02:01.49922+00  |
```

```

File: _posts/2023-04-30-openai-supabase-summary.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    return response.choices[0].message.content
+```
+
+The `gpt_4_summarizer` step is a custom ZenML step that leverages GPT-4 to generate summaries of the input data. It takes `SummarizerParams` and a list of documents as input, along with the `StepContext` to access previous pipeline runs.
+
+The step first retrieves the secret API key for OpenAI and checks for any previous analysis. If a previous analysis is found, it is included in the input to GPT-4. This allows GPT-4 to consider the previous analysis while generating the new summary.
+
+The step then calls the GPT-4 API with the specified parameters and input data, including the documents and the previous analysis. The API returns a summary, which is then returned as the output of the step.
+
+This custom step can be adapted for other use cases by modifying the input parameters, the GPT-4 model, or the API call. For example, you could use this step to generate summaries for different types of data, use a different language model, or customize the API call to suit your specific needs. By customizing this step, you can harness the power of GPT-4 and Zen
```
Comment (strickvl) [inline _posts/2023-04-30-openai-supabase-summary.md:148]:
```suggestion
This custom step can be adapted for other use cases by modifying the input parameters, the GPT-4 model, or the API call. For example, you could use this step to generate summaries for different types of data, use a different language model, or customize the API call to suit your specific needs. By customizing this step, you can harness the power of GPT-4 and ZenML.
```

Given the cost difference, you'd probably be better off using `chat-3.5-turbo` instead. It's more than capable enough to handle this task.

File: _posts/2023-04-30-openai-supabase-summary.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+D --> E[Versioned Summaries]
+E --> F[GPT-4 Comparison]
+F --> G[Insights]
+```
+
+**Figure 4:** System Architecture Diagram
+
+## 🎯 Conclusion
+
+This case study demonstrates the power of combining ZenML with GPT-4 to analyze and version data from a Supabase database. By using ZenML to manage data versioning and GPT-4 for analysis, we can overcome the limitations of LLMs and gain valuable insights from our data. If you're interested in leveraging the latest technology for your own projects, consider using ZenML in conjunction with GPT-4 and Supabase to unlock the full potential of your data.
```
Comment (strickvl) [inline _posts/2023-04-30-openai-supabase-summary.md:175]: Probably wants some links or something back to ZenML, maybe to other LLM- or OpenAI- related posts and docs etc? Also to the project repo itself?

**Review by strickvl — CHANGES_REQUESTED** [review]
Interesting project. I wonder if there's a way to characterise what you did beyond the description. I.e. does it fall into a pattern or common grouping? Wonder whether it makes sense to use the word MLOps as a characterisation.

**Comment by htahir1** [comment]
@strickvl Thanks for the review! I think the power in this project is the ability to read from an "enterprise" dataset and summarize it over time. You can apply this over time for many things (e.g. I could combine a user table with the summaries here and have the last step of the pipeline automatically generated emails to send personalized marketing to people). There is a lot of power in having a snapshot of recent data.

Im looking for suggestions how to generalize this. Any ideas appreciated!

**Comment by strickvl** [comment]
@htahir1 Yeah maybe that just needs to be emphasised a bit more. Somehow I also wondered whether the data collation could be made a bit more fancy (instead of passing a list of titles, compress them down into vector space and that way you might get to pass more titles down). From what I read, one of the big concerns people have at the moment is both cost and passing their data to OpenAI. If you're compressing things down as embeddings, you might at least partially address both of those. (Maybe something for a different project, though I think it would be useful at some point for ZenML to address these needs that people have, or at the very least tip our hats at those needs).

**Comment by htahir1** [comment]
@strickvl Im not sure embeddings works in this use-case though? I need it to summarize/analyze the text, and embeddings AFAIK dont mantain semantic meaning after compression right. I could be wrong, is there a way to do this? Would be much cooler

**Comment by strickvl** [comment]
We're not really doing anything fancy, just summarising topics, so I'm not sure we need all the precise semantic / syntactic sugar on top of that. I'm sure it must be possible. [This pattern](https://www.piratewires.com/p/compression-prompts-gpt-hidden-dialects) is what I was thinking of somehow, but it defeats the point to use OpenAI to make the summarisation... (Tried this briefly and it didn't work.)

Maybe something extra to throw in there would just be to get / generate the embeddings for the input text in a separate step, do that every time and then just get a cosine similarity between the embeddings (i.e. locally) as a way of deciding whether to then pass the data to GPT-4 in the first place. I.e. if the topics are similar (above a certain threshold, then no need to do the summarisation again, but just post to the alerter that no big changes detected).

https://colab.research.google.com/drive/1NR_Exqs_1eMtBNWe_qXQHenSbJBU5bzn?usp=sharing does what I was suggesting above. Goes in the direction of reducing a user's OpenAI bill. But maybe too complicated / inapplicable to the use case?

File: _posts/2023-04-30-openai-supabase-summary.md [hunk 7]
```diff
@@ -170,6 +184,18 @@ F --> G[Insights]
 
 **Figure 4:** System Architecture Diagram
 
+## 🌟 The Power of Summarizing Enterprise Data Over Time
+
+The true power of this project lies in its ability to read from an enterprise dataset and generate summaries over time. By combining the capabilities of ZenML, GPT-4, and Supabase, we can create a powerful system that can be applied to a wide range of use cases.
+
+For example, imagine combining a user table with the summaries generated by this pipeline. The last step of the pipeline could automatically generate personalized marketing emails based on the summarized data, allowing for highly targeted and effective communication with customers.
```
Comment (strickvl) [inline _posts/2023-04-30-openai-supabase-summary.md:191]: Perhaps just me, but the 'vibes' of this use case don't really resonate. Is there a way to rephrase or reframe this?

File: _posts/2023-04-30-openai-supabase-summary.md [hunk 8]
```diff
@@ -170,6 +184,18 @@ F --> G[Insights]
...
+
+Another potential application is monitoring and analyzing trends in your data over time. By generating summaries at regular intervals, you can gain valuable insights into how your data is changing and evolving, allowing you to make more informed decisions and adapt your strategies accordingly.
```
Comment (strickvl) [inline _posts/2023-04-30-openai-supabase-summary.md:193]: Don't find this point very convincing. We have lots of tools for doing this already. Maybe the thing to say here is that it allows for quick POCs, or for implementations from people with less technical experience? or it allows for quick iteration and experimentation on the end-to-end use case which people can then adapt with more built-to-purpose tools?

**Comment by htahir1** [comment]
@strickvl Really love your suggestion but id first launch this and then iterate. Ive left other thoughts in place

File: _posts/2023-04-30-openai-supabase-summary.md (hunk 7 above)
Comment (htahir1) [reply _posts/2023-04-30-openai-supabase-summary.md:191]: I've rephrased it in the new version, let me know if its better vibes? :-)

File: _posts/2023-04-30-openai-supabase-summary.md (hunk 8 above)
Comment (htahir1) [reply _posts/2023-04-30-openai-supabase-summary.md:193]: I understand. I think maybe we should rather focus on ZenML benefits like getting into production ready orchestrators etc. I'll add that

File: _posts/2023-04-30-openai-supabase-summary.md (hunk 7 above)
Comment (strickvl) [reply _posts/2023-04-30-openai-supabase-summary.md:191]: What new version?

Comment (htahir1) [reply _posts/2023-04-30-openai-supabase-summary.md]: @strickvl just pushed

**Review by strickvl — APPROVED** [review]
Nice

File: _posts/2023-04-30-openai-supabase-summary.md [hunk 9]
```diff
@@ -0,0 +1,198 @@
+---
+layout: post
+author: "Hamza Tahir"
+title: "Using ZenML with LLMs to Analyze and Version Data from Databases: A Case Study with you-tldr.com and Supabase/GPT-4"
```
Comment (htahir1) [inline _posts/2023-04-30-openai-supabase-summary.md:4]:
```suggestion
title: "Using ZenML with LLMs to Analyze Your Databases: A Case Study with you-tldr.com and Supabase/GPT-4"
```

**Pull request merged**